src/main/java/com/englishtown/bitbucket/hook/MirrorBucketProcessor.java

```diff
@@ -0,0 +1,147 @@
+
+    private final I18nService i18nService;
+    private final PasswordEncryptor passwordEncryptor;
+    private final RepositoryService repositoryService;
+    private final ScmService scmService;
+    private final SecurityService securityService;
+    private final Duration timeout;
+
+    public MirrorBucketProcessor(I18nService i18nService, PasswordEncryptor passwordEncryptor,
+                                 ApplicationPropertiesService propertiesService, RepositoryService repositoryService,
+                                 ScmService scmService, SecurityService securityService) {
+        this.i18nService = i18nService;
+        this.passwordEncryptor = passwordEncryptor;
+        this.repositoryService = repositoryService;
+        this.scmService = scmService;
+        this.securityService = securityService;
+
+        timeout = Duration.ofSeconds(propertiesService.getPluginProperty(PROP_TIMEOUT, 120L));
+    }
+
+    @Override
+    public void process(@Nonnull String key, @Nonnull List<MirrorRequest> requests) {
+        if (requests.isEmpty()) {
+            return;
+        }
+        // Every request is for the same mirror URL, and the same repository ID. In case the
+        // settings (e.g. username/password) have been changed since the first request was
+        // queued, we process the _last_ request in the list. Since mirroring pushes all of
+        // the configured refspecs, any single request should roll up changes from any number
+        // of requests
+        MirrorRequest request = requests.get(requests.size() - 1);
+
+        securityService.withPermission(Permission.REPO_READ, "Mirror changes")
+                .call(() -> {
+                    Repository repository = repositoryService.getById(request.getRepositoryId());
+                    if (repository == null) {
+                        log.debug("{}: Repository has been deleted", request.getRepositoryId());
+                        return null;
+                    }
+                    if (repositoryService.isEmpty(repository)) {
+                        log.debug("{}: The repository is empty", repository);
+                        return null;
+                    }
+                    runMirrorCommand(request.getSettings(), repository);
+
+                    return null;
+                });
+    }
+
+    private void runMirrorCommand(MirrorSettings settings, Repository repository) {
+        log.warn("{}: Preparing to push changes to mirror", repository);
+
+        String password = passwordEncryptor.decrypt(settings.password);
+        String authenticatedUrl = getAuthenticatedUrl(settings.mirrorRepoUrl, settings.username, password);
+
+        // Call push command with the prune flag and refspecs for heads and tags
+        // Do not use the mirror flag as pull-request refs are included
+        ScmCommandBuilder<?> builder = scmService.createBuilder(repository)
```

Comment:
There's no need to cast this to a `GitScmCommandBuilder`, because the type-safe `GitPushBuilder` isn't being used anyway.